File: day2/main.py
"""
https://adventofcode.com/2020/day/2
"""
FILE = 'input.txt'

class Policy:
    def __init__(self, letter, lower, upper):
        self.letter = letter
        self.lower = lower
        self.upper = upper

    def test1(self, password):
        return self.lower <= password.count(self.letter) <= self.upper

    def test2(self, password):
        # Take logical XOR
        return (password[self.lower - 1] == self.letter) != (password[self.upper - 1] == self.letter)

def parse_input(line):
    """
    Parses an input line and produces a policy and password
    """
    limits, letter, password = line.split()
    lower, upper = limits.split('-')
    policy = Policy(letter[0], int(lower), int(upper))
    return policy, password

def main():
    num_valid_1 = 0
    num_valid_2 = 0
    with open(FILE) as f:
        for line in f:
            policy, password = parse_input(line)
            if policy.test1(password):
                num_valid_1 += 1
            if policy.test2(password):
                num_valid_2 += 1
    print(f'Part 1: {num_valid_1}')
    print(f'Part 2: {num_valid_2}')

if __name__ == '__main__':
    main()

File: day2/main.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <utility>

#include "Policy.h"

using std::cout;
using std::pair;
using std::string;
using std::vector;

vector<pair<Policy, string>> readFile(string filename);
int countTest1(vector<pair<Policy, string>> policyPasswords);
int countTest2(vector<pair<Policy, string>> policyPasswords);

int main(int argc, char **argv)
{
    vector<pair<Policy, string>> policyPasswords = readFile("input.txt");
    int count1 = 0;
    int count2 = 0;
    for (pair<Policy, string> policyPassword : policyPasswords)
    {
        if (std::get<0>(policyPassword).test1(std::get<1>(policyPassword)))
        {
            count1++;
        }
        if (std::get<0>(policyPassword).test2(std::get<1>(policyPassword)))
        {
            count2++;
        }
    }
    cout << "Part 1: " << count1 << "\n";
    cout << "Part 1: " << count2 << "\n";
    return 0;
}

vector<pair<Policy, string>> readFile(string filename)
{
    std::ifstream file(filename);
    string token;
    vector<pair<Policy, string>> policyPasswords;
    while (getline(file, token, '-'))
    {
        int lower = stoi(token);
        getline(file, token, ' ');
        int upper = stoi(token);
        getline(file, token, ':');
        char letter = token.at(0);
        getline(file, token, ' ');
        getline(file, token, '\n');
        string password = token;
        Policy policy(lower, upper, letter);
        pair<Policy, string> policyPassword(policy, password);
        policyPasswords.push_back(policyPassword);
    }
    return policyPasswords;
}
